packages/fields-document/src/DocumentEditor/utils-model.ts

```diff
@@ -0,0 +1,155 @@
+import {
+  Editor,
+  Node,
+  type NodeEntry,
+  Path,
+  Transforms,
+  Element,
+  type PathRef,
+  Text,
+  type Location,
+  type Point,
+} from 'slate'
+// import { ReactEditor } from 'slate-react' // TODO
+import { type ElementFromValidation } from '../structure-validation'
+
+export type Mark =
+  | 'bold'
+  | 'italic'
+  | 'underline'
+  | 'strikethrough'
+  | 'code'
+  | 'superscript'
+  | 'subscript'
+  | 'keyboard'
+
+export const allMarks: Mark[] = [
+  'bold',
+  'italic',
+  'underline',
+  'strikethrough',
+  'code',
+  'superscript',
+  'subscript',
+  'keyboard',
+]
+
+export const isElementActive = (
+  editor: Editor,
+  format: Exclude<ElementFromValidation, { text: string }>['type']
+) => {
+  const [match] = Editor.nodes(editor, {
+    match: n => n.type === format,
+  })
+
+  return !!match
+}
+
+export function clearFormatting(editor: Editor) {
+  Transforms.unwrapNodes(editor, {
+    match: node => node.type === 'heading' || node.type === 'blockquote' || node.type === 'code',
+  })
+  Transforms.unsetNodes(editor, allMarks, { match: Text.isText })
+}
+
+export function moveChildren(
+  editor: Editor,
+  parent: NodeEntry | Path,
+  to: Path,
+  shouldMoveNode: (node: Node) => boolean = () => true
+) {
+  const parentPath = Path.isPath(parent) ? parent : parent[1]
+  const parentNode = Path.isPath(parent) ? Node.get(editor, parentPath) : parent[0]
+  if (!(Element.isElement(parentNode) && Editor.isBlock(editor, parentNode))) return
+
+  for (let i = parentNode.children.length - 1; i >= 0; i--) {
+    if (shouldMoveNode(parentNode.children[i])) {
+      const childPath = [...parentPath, i]
+      Transforms.moveNodes(editor, { at: childPath, to })
+    }
+  }
+}
+
+export function insertNodesButReplaceIfSelectionIsAtEmptyParagraphOrHeading(
+  editor: Editor,
+  nodes: Node | Node[]
+) {
+  let pathRefForEmptyNodeAtCursor: PathRef | undefined
+  const entry = Editor.above(editor, {
+    match: node => node.type === 'heading' || node.type === 'paragraph',
+  })
+  if (entry && Node.string(entry[0]) === '') {
+    pathRefForEmptyNodeAtCursor = Editor.pathRef(editor, entry[1])
+  }
+  Transforms.insertNodes(editor, nodes)
+  let path = pathRefForEmptyNodeAtCursor?.unref()
+  if (path) {
+    Transforms.removeNodes(editor, { at: path })
+    // even though the selection is in the right place after the removeNodes
+    // for some reason the editor blurs so we need to focus it again
+
+    // TODO: this causes 'TypeError: ([...].createContext) is not a function' in SSR
+    // not sure how to reproduce state in which this is neccessary
+    // ReactEditor.focus(editor)
```

Comment:
As stated in the description, this is the only functional change. Importing anything from `slate-react` causes
```

 ⨯ TypeError: (0 , react__WEBPACK_IMPORTED_MODULE_3__.createContext) is not a function
    at eval (webpack-internal:///(rsc)/../../node_modules/.pnpm/slate-react@0.102.0_react-dom@18.2.0_react@18.2.0_slate@0.102.0/node_modules/slate-react/dist/index.es.js:108:91)
    at (rsc)/../../node_modules/.pnpm/slate-react@0.102.0_react-dom@18.2.0_react@18.2.0_slate@0.102.0/node_modules/slate-react/dist/index.es.js (/home/exar/os/keystone/examples/framework-nextjs-app-directory/.next/server/vendor-chunks/slate-react@0.102.0_react-dom@18.2.0_react@18.2.0_slate@0.102.0.js:20:1)
```
because it uses `createContext`. I alleviated this problem in other place by having two versions of `createDocumentEditor`:  `packages/fields-document/src/DocumentEditor/editor-create-display.ts` and `packages/fields-document/src/DocumentEditor/editor-create-validation.ts`.

However in this case I am kind of clueless what the original problem was and how to workaround it.

Reply:
When did this happen for you?

Reply:
I have resolved this with b85894f, by using a dependency passing pattern to side-step the problem without duplicating the code.

Reply:
The problem @marekryb is that `ReactEditor` has a call to `createContext`, which isn't allowed on the server.